eZ/Bundle/EzPublishCoreBundle/Entity/EntityManagerFactory.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Bundle\EzPublishCoreBundle\Entity;
+
+use Doctrine\ORM\EntityManagerInterface;
+use eZ\Bundle\EzPublishCoreBundle\ApiLoader\RepositoryConfigurationProvider;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+class EntityManagerFactory
+{
+    /** @var \eZ\Bundle\EzPublishCoreBundle\ApiLoader\RepositoryConfigurationProvider */
+    private $repositoryConfigurationProvider;
+
+    /** @var \Symfony\Component\DependencyInjection\ContainerInterface */
+    private $container;
+
+    public function __construct(
+        RepositoryConfigurationProvider $repositoryConfigurationProvider,
+        ContainerInterface $container
+    ) {
+        $this->repositoryConfigurationProvider = $repositoryConfigurationProvider;
+        $this->container = $container;
+    }
+
+    public function getEntityManager(): EntityManagerInterface
+    {
+        $repositoryConfig = $this->repositoryConfigurationProvider->getRepositoryConfig();
+
+        if (isset($repositoryConfig['storage']['connection'])) {
+            $entityManagerId = $this->getEntityManagerServiceId($repositoryConfig['storage']['connection']);
+        } else {
+            $defaultEntityManagerId = $this->getEntityManagerServiceId($this->container->getParameter('doctrine.default_connection'));
+            $entityManagerId = $this->container->has($defaultEntityManagerId)
+                ? $defaultEntityManagerId
+                : 'doctrine.orm.entity_manager';
+        }
+
+        if (!$this->container->has($entityManagerId)) {
+            throw new \InvalidArgumentException(
+                "Invalid Doctrine Entity Manager '{$entityManagerId}' for Repository '{$repositoryConfig['alias']}'." .
+                'Valid Entity Managers are: ' . implode(', ', array_keys($this->container->getParameter('doctrine.entity_managers')))
+            );
+        }
+
+        return $this->container->get($entityManagerId);
```

Comment:
would it be possible to use service locator for this (so manipulating Definition creation so it's passed properly to service locator smaller container)? Or does Doctrine require EM to be public?

Reply:
It's not going to work. It would require to compose a list of our Entity Managers statically but we can't do that.

Reply:
We definitely cannot rely on main application container, so we have to find a workaround. 

I also think it's doable. Generating a Service Locator during the compiler pass is something that can be done. See the following classes for possible inspiration:
`\Symfony\Component\DependencyInjection\Compiler\ServiceLocatorTagPass`
`\Symfony\Component\Serializer\DependencyInjection\SerializerPass`
`\JMS\SerializerBundle\DependencyInjection\Compiler\DoctrinePass`

For documentation on service locators defined during compilation: https://symfony.com/doc/current/service_container/service_subscribers_locators.html#using-service-locators-in-compiler-passes

Adding dependency on main application container and the fact that entity managers need to be public is major debt in my opinion and should not be left as is. I understand it's bothersome, and I'm sorry about that.

Reply:
I can try but having public Entity Managers it's not something I invented out of conveniency - this is how Doctrine does it. We also rely on retrieving Connection services directly from the container in `StorageConnectionFactory` which was the main inspiration for `EntityManagerFactory`. I will see if utilizing service locators is possible :+1:

Reply:
Please check f55fb89. I reimplemented it using `ServiceLocator` and a compiler pass to build the list of locatable Entity Managers.